README.md
## ThatcherTiler

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/10407788/231709578-04c9ae59-c264-4319-be9d-a70d1bd98a1f.jpg"/>
  <p align="center">ThatcherTiler: expect some features to be dropped.</p>
</p>


---

**Documentation**:

**Source Code**: <a href="https://github.com/developmentseed/thatchertiler" target="_blank">https://github.com/developmentseed/thatchertiler</a>

---

`ThatcherTiler`* is a lightweight Raster/Vector tiles server based on the great [**PM**Tiles](https://github.com/protomaps/PMTiles) *Cloud-optimized + compressed single-file tile archives for vector and raster maps*.

While the original idea behind *PMTiles* is to create a single file which can be accessed directly from map client (e.g https://protomaps.com/docs/frontends/maplibre) some map client yet do not support it (e.g `Mapbox`) while a need for a `tile server` between the map client and the archive.

`ThatcherTiler` is a **DEMO** project to showcase the use of [aiopmtiles](https://github.com/developmentseed/aiopmtiles).


## Install

```bash
python -m pip install pip -U
git clone https://github.com/developmentseed/thatchertiler.git
cd thatchertiler
python -m pip install -e ".[server]"
```

## Launch

```bash
uvicorn thatchertiler.main:app --port 8080 --reload

>> INFO:     Uvicorn running on http://127.0.0.1:8080 (Press CTRL+C to quit)
```

`http://127.0.0.1:8080/api.html`

![](https://user-images.githubusercontent.com/10407788/231717328-6d0608fa-145b-480e-ba48-629b0e4e3e97.png)


`http://127.0.0.1:8080/map?url=https://protomaps.github.io/PMTiles/protomaps(vector)ODbL_firenze.pmtiles`

![](https://user-images.githubusercontent.com/10407788/231720737-87cf14a9-35b3-4451-a183-37b80dc16e93.png)

## License

See [LICENSE](https://github.com/developmentseed/thatchertiler/blob/main/LICENSE)

## Authors

Created by [Development Seed](<http://developmentseed.org>)

*`ThatcherTiler` name comes from a long a intense debate among @nerik, @kamicut, @batpad and @vincentsarago to find the better of the worse name. `Thatcher` is a reference to [`Margaret Thatcher`](https://fr.wikipedia.org/wiki/Margaret_Thatcher) which was once **P**rime **M**inister of the United Kingdom.


CHANGES.md

## 0.1.0 (TBD)

* Initial release.


tests/__init__.py
"""thatchertiler."""


tests/conftest.py
"""``pytest`` configuration."""

import os

import pytest
from starlette.testclient import TestClient


@pytest.fixture
def set_env(monkeypatch):
    """Set Env variables."""
    monkeypatch.setenv("AWS_ACCESS_KEY_ID", "jqt")
    monkeypatch.setenv("AWS_SECRET_ACCESS_KEY", "rde")
    monkeypatch.setenv("AWS_DEFAULT_REGION", "us-west-2")
    monkeypatch.setenv("AWS_REGION", "us-west-2")
    monkeypatch.delenv("AWS_PROFILE", raising=False)
    monkeypatch.setenv("AWS_CONFIG_FILE", "/tmp/noconfigheere")
    monkeypatch.setenv("THATCHERTILER_API_CACHECONTROL", "private, max-age=3600")


@pytest.fixture(autouse=True)
def app(set_env) -> TestClient:
    """Create App."""
    from thatchertiler.main import app

    return TestClient(app)


@pytest.fixture(autouse=True)
def data_dir() -> str:
    """fixture directory."""
    return os.path.join(os.path.dirname(__file__), "fixtures")


tests/test_main.py
"""Test thatchertiler.main.app."""


def test_health(app):
    """Test /healthz endpoint."""
    response = app.get("/healthz")
    assert response.status_code == 200
    assert response.json() == {"ping": "pong!"}

    response = app.get("/api")
    assert response.status_code == 200

    response = app.get("/api.html")
    assert response.status_code == 200


thatchertiler/model.py
"""response models."""

from typing import Dict, List, Literal, Optional, Tuple

from pydantic import BaseModel, Field, model_validator


class LayerJSON(BaseModel):
    """
    https://github.com/mapbox/tilejson-spec/tree/master/3.0.0#33-vector_layers
    """

    id: str
    fields: Dict = Field(default_factory=dict)
    description: Optional[str] = None
    minzoom: Optional[int] = None
    maxzoom: Optional[int] = None


class TileJSON(BaseModel):
    """TileJSON model."""

    tilejson: str = "3.0.0"
    name: Optional[str] = None
    description: Optional[str] = None
    version: str = "1.0.0"
    attribution: Optional[str] = None
    template: Optional[str] = None
    legend: Optional[str] = None
    scheme: Literal["xyz", "tms"] = "xyz"
    tiles: List[str] = None
    vector_layers: Optional[List[LayerJSON]] = None
    grids: Optional[List[str]] = None
    data: Optional[List[str]] = None
    minzoom: int = Field(0, ge=0, le=30)
    maxzoom: int = Field(30, ge=0, le=30)
    fillzoom: Optional[int] = None
    bounds: List[float] = [180, -85.05112877980659, 180, 85.0511287798066]
    center: Optional[Tuple[float, float, int]] = None

    @model_validator(mode="after")
    def compute_center(self):
        """Compute center if it does not exist."""
        bounds = self.bounds
        if not self.center:
            self.center = (
                (bounds[0] + bounds[2]) / 2,
                (bounds[1] + bounds[3]) / 2,
                self.minzoom,
            )
        return self


class StyleJSON(BaseModel):
    """
    Simple Mapbox/Maplibre Style JSON model.

    Based on https://docs.mapbox.com/help/glossary/style/

    """

    version: int = 8
    name: Optional[str] = None
    metadata: Optional[Dict] = None
    layers: List[Dict]
    sources: Dict
    center: List[float] = None
    zoom: int


thatchertiler/factory.py
"""thatchertiler.factory."""

import abc
import json
import pathlib
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Dict, List, Literal, Optional, Tuple, Type, Union
from urllib.parse import urlencode

import jinja2
from aiopmtiles import Reader
from fastapi import APIRouter, Body, Depends, Path, Query
from fastapi.dependencies.utils import get_parameterless_sub_dependant
from fastapi.params import Depends as DependsFunc
from pmtiles.tile import Compression
from starlette.requests import Request
from starlette.responses import HTMLResponse, Response
from starlette.routing import Match, compile_path, replace_params
from starlette.templating import Jinja2Templates
from typing_extensions import Annotated

from thatchertiler.model import StyleJSON, TileJSON
from thatchertiler.routing import EndpointScope

DEFAULT_TEMPLATES = Jinja2Templates(
    directory="",
    loader=jinja2.ChoiceLoader([jinja2.PackageLoader(__package__, "templates")]),
)

tms_data = pathlib.Path(__file__).parent.joinpath("data", "WebMercatorQuad.json")


class MediaType(str, Enum):
    """Responses Media types formerly known as MIME types."""

    unknown = "application/x-binary"
    avif = "image/avif"
    tif = "image/tiff; application=geotiff"
    jp2 = "image/jp2"
    png = "image/png"
    pngraw = "image/png"
    jpeg = "image/jpeg"
    jpg = "image/jpg"
    webp = "image/webp"
    npy = "application/x-binary"
    xml = "application/xml"
    json = "application/json"
    geojson = "application/geo+json"
    html = "text/html"
    text = "text/plain"
    pbf = "application/x-protobuf"
    mvt = "application/x-protobuf"


class XMLResponse(Response):
    """XML Response"""

    media_type = "application/xml"


@dataclass  # type: ignore
class FactoryExtension(metaclass=abc.ABCMeta):
    """Factory Extension."""

    @abc.abstractmethod
    def register(self, factory: "TilerFactory"):
        """Register extension to the factory."""
        ...


def PathParams(url: Annotated[str, Query(description="PMTiles archive URL.")]) -> str:
    """Dataset Parameter"""
    return url


@dataclass
class TilerFactory:
    """thatchertiler endpoint factory."""

    path_dependency: Callable[..., str] = PathParams

    # FastAPI router
    router: APIRouter = field(default_factory=APIRouter)

    # Router Prefix is needed to find the path for /tile if the TilerFactory.router is mounted
    # with other router (multiple `.../tile` routes).
    # e.g if you mount the route with `/cog` prefix, set router_prefix to cog and
    router_prefix: str = ""

    # add dependencies to specific routes
    route_dependencies: List[Tuple[List[EndpointScope], List[DependsFunc]]] = field(
        default_factory=list
    )

    extensions: List[FactoryExtension] = field(default_factory=list)

    templates: Jinja2Templates = DEFAULT_TEMPLATES

    with_wmts: bool = False
    with_map: bool = False

    def __post_init__(self):
        """Post Init: register route and configure specific options."""
        # Register endpoints
        self.register_routes()

        # Register Extensions
        for ext in self.extensions:
            ext.register(self)

        # Update endpoints dependencies
        for scopes, dependencies in self.route_dependencies:
            self.add_route_dependencies(scopes=scopes, dependencies=dependencies)

    def url_for(self, request: Request, name: str, **path_params: Any) -> str:
        """Return full url (with prefix) for a specific endpoint."""
        url_path = self.router.url_path_for(name, **path_params)
        base_url = str(request.base_url)
        if self.router_prefix:
            prefix = self.router_prefix.lstrip("/")
            # If we have prefix with custom path param we check and replace them with
            # the path params provided
            if "{" in prefix:
                _, path_format, param_convertors = compile_path(prefix)
                prefix, _ = replace_params(
                    path_format, param_convertors, request.path_params.copy()
                )
            base_url += prefix

        return str(url_path.make_absolute_url(base_url=base_url))

    def add_route_dependencies(
        self,
        *,
        scopes: List[EndpointScope],
        dependencies=List[DependsFunc],
    ):
        """Add dependencies to routes.

        Allows a developer to add dependencies to a route after the route has been defined.

        """
        for route in self.router.routes:
            for scope in scopes:
                match, _ = route.matches({"type": "http", **scope})  # type: ignore
                if match != Match.FULL:
                    continue

                # Mimicking how APIRoute handles dependencies:
                # https://github.com/tiangolo/fastapi/blob/1760da0efa55585c19835d81afa8ca386036c325/fastapi/routing.py#L408-L412
                for depends in dependencies[::-1]:
                    route.dependant.dependencies.insert(  # type: ignore
                        0,
                        get_parameterless_sub_dependant(
                            depends=depends, path=route.path_format  # type: ignore
                        ),
                    )

                # Register dependencies directly on route so that they aren't ignored if
                # the routes are later associated with an app (e.g. app.include_router(router))
                # https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/applications.py#L337-L360
                # https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/routing.py#L677-L678
                route.dependencies.extend(dependencies)  # type: ignore

    def register_routes(self):
        """Register Tiler Routes."""

        self._register_metadata()
        self._register_tiles()
        self._register_style()

        if self.with_map:
            self._register_map()

        if self.with_wmts:
            self._register_wmts()

    def _register_metadata(self):
        """register /metadata endpoint."""

        @self.router.get("/metadata")
        async def metadata(dataset=Depends(self.path_dependency)):
            """get PMTiles Metadata."""
            async with Reader(dataset) as src:
                return await src.metadata()

    def _register_tiles(self):
        """register /tiles and /tilejson.json endpoints."""

        @self.router.get("/tiles/{z}/{x}/{y}", response_class=Response)
        async def tiles_endpoint(
            z: Annotated[int, Path(ge=0, le=30, description="TMS tiles's zoom level")],
            x: Annotated[int, Path(description="TMS tiles's column")],
            y: Annotated[int, Path(description="TMS tiles's row")],
            dataset=Depends(self.path_dependency),
        ):
            """get Tile."""
            headers: Dict[str, str] = {}

            async with Reader(dataset) as src:
                data = await src.get_tile(z, x, y)
                if src.tile_compression != Compression.NONE:
                    headers["Content-Encoding"] = src.tile_compression.name.lower()

                return Response(
                    data,
                    media_type=MediaType[src.tile_type.name.lower()].value,
                    headers=headers,
                )

        @self.router.get(
            "/tilejson.json",
            response_model=TileJSON,
            response_model_exclude_none=True,
        )
        async def tilejson_endpoint(
            request: Request, dataset=Depends(self.path_dependency)
        ):
            """get TileJSON."""
            async with Reader(dataset) as src:
                meta = await src.metadata()

                route_params = {
                    "z": "{z}",
                    "x": "{x}",
                    "y": "{y}",
                }
                tiles_url = self.url_for(request, "tiles_endpoint", **route_params)

                if request.query_params._list:
                    tiles_url += f"?{urlencode(request.query_params._list)}"

                tilejson = {
                    "name": "pmtiles",
                    "version": "1.0.0",
                    "scheme": "xyz",
                    "tiles": [tiles_url],
                    "minzoom": src.minzoom,
                    "maxzoom": src.maxzoom,
                    "bounds": src.bounds,
                    "center": src.center,
                }

                if vector_layers := meta.get("vector_layers"):
                    tilejson["vector_layers"] = vector_layers

            return tilejson

    def _register_style(self):
        """register /style.json endpoint."""

        @self.router.get(
            "/style.json",
            response_model=StyleJSON,
            response_model_exclude_none=True,
        )
        async def stylejson_endpoint(
            request: Request, dataset=Depends(self.path_dependency)
        ):
            """get StyleJSON."""
            route_params = {
                "z": "{z}",
                "x": "{x}",
                "y": "{y}",
            }
            tiles_url = self.url_for(request, "tiles_endpoint", **route_params)

            if request.query_params._list:
                tiles_url += f"?{urlencode(request.query_params._list)}"

            style_json: Dict[str, Any]
            async with Reader(dataset) as src:
                if src.is_vector:
                    style_json = {
                        "sources": {
                            "pmtiles": {
                                "type": "vector",
                                "scheme": "xyz",
                                "tiles": [tiles_url],
                                "minzoom": src.minzoom,
                                "maxzoom": src.maxzoom,
                                "bounds": src.bounds,
                            },
                            "basemap": {
                                "type": "raster",
                                "tiles": [
                                    "https://tile.openstreetmap.org/{z}/{x}/{y}.png",
                                ],
                                "tileSize": 256,
                                "attribution": '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
                            },
                        },
                    }

                    layers: List[Dict[str, Any]] = [
                        {
                            "id": "basemap",
                            "type": "raster",
                            "source": "basemap",
                            "minzoom": 0,
                            "maxzoom": 20,
                        }
                    ]

                    meta = await src.metadata()
                    if vector_layers := meta.get("vector_layers"):
                        for layer in vector_layers:
                            layer_id = layer["id"]
                            if layer_id == "mask":
                                layers.append(
                                    {
                                        "id": f"{layer_id}_fill",
                                        "type": "fill",
                                        "source": "pmtiles",
                                        "source-layer": layer_id,
                                        "filter": ["==", ["geometry-type"], "Polygon"],
                                        "paint": {
                                            "fill-color": "black",
                                            "fill-opacity": 0.8,
                                        },
                                    }
                                )

                            else:
                                layers.append(
                                    {
                                        "id": f"{layer_id}_fill",
                                        "type": "fill",
                                        "source": "pmtiles",
                                        "source-layer": layer_id,
                                        "filter": ["==", ["geometry-type"], "Polygon"],
                                        "paint": {
                                            "fill-color": "rgba(200, 100, 240, 0.4)",
                                            "fill-outline-color": "#000",
                                        },
                                    }
                                )

                            layers.append(
                                {
                                    "id": f"{layer_id}_stroke",
                                    "source": "pmtiles",
                                    "source-layer": layer_id,
                                    "type": "line",
                                    "filter": ["==", ["geometry-type"], "LineString"],
                                    "paint": {
                                        "line-color": "#000",
                                        "line-width": 1,
                                        "line-opacity": 0.75,
                                    },
                                }
                            )
                            layers.append(
                                {
                                    "id": f"{layer_id}_point",
                                    "source": "pmtiles",
                                    "source-layer": layer_id,
                                    "type": "circle",
                                    "filter": ["==", ["geometry-type"], "Point"],
                                    "paint": {
                                        "circle-color": "#000",
                                        "circle-radius": 2.5,
                                        "circle-opacity": 0.75,
                                    },
                                }
                            )

                    style_json["layers"] = layers

                else:
                    style_json = {
                        "sources": {
                            "pmtiles": {
                                "type": "raster",
                                "scheme": "xyz",
                                "tiles": [tiles_url],
                                "minzoom": src.minzoom,
                                "maxzoom": src.maxzoom,
                                "bounds": src.bounds,
                            },
                            "basemap": {
                                "type": "raster",
                                "tiles": [
                                    "https://tile.openstreetmap.org/{z}/{x}/{y}.png",
                                ],
                                "tileSize": 256,
                                "attribution": '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
                            },
                        },
                        "layers": [
                            {
                                "id": "basemap",
                                "type": "raster",
                                "source": "basemap",
                                "minzoom": 0,
                                "maxzoom": 20,
                            },
                            {
                                "id": "raster",
                                "type": "raster",
                                "source": "pmtiles",
                            },
                        ],
                    }
                style_json["center"] = [src.center[0], src.center[1]]
                style_json["zoom"] = src.center[2]

            return style_json

    def _register_map(self):
        """add /map endpoint."""

        @self.router.get("/map", response_class=HTMLResponse)
        async def map_endpoint(request: Request, dataset=Depends(self.path_dependency)):
            """Handle /index.html."""
            stylejson_url = self.url_for(request, "stylejson_endpoint")
            if request.query_params._list:
                stylejson_url += f"?{urlencode(request.query_params._list)}"

            return self.templates.TemplateResponse(
                name="index.html",
                context={
                    "request": request,
                    "style_endpoint": stylejson_url,
                },
                media_type="text/html",
            )

    def _register_wmts(self):
        """add /wmts endpoint."""

        @self.router.get("/WMTSCapabilities.xml", response_class=XMLResponse)
        async def wmts(request: Request, dataset=Depends(self.path_dependency)):
            """OGC WMTS endpoint."""
            async with Reader(dataset) as src:
                route_params = {
                    "z": "{TileMatrix}",
                    "x": "{TileCol}",
                    "y": "{TileRow}",
                }
                tiles_url = self.url_for(request, "tiles_endpoint", **route_params)

                qs_key_to_remove = [
                    "service",
                    "request",
                ]
                qs = [
                    (key, value)
                    for (key, value) in request.query_params._list
                    if key.lower() not in qs_key_to_remove
                ]
                if qs:
                    tiles_url += f"?{urlencode(qs)}"

                with open(tms_data, "r") as tms:
                    tile_matrix_set = json.loads(tms.read())

                zooms = list(map(str, range(src.minzoom, src.maxzoom + 1)))
                matrices = filter(
                    lambda matrix: matrix["id"] in zooms,
                    tile_matrix_set["tileMatrices"],
                )

                tileMatrix = []
                for matrix in list(matrices):
                    tm = f"""<TileMatrix>
                        <ows:Identifier>{matrix['id']}</ows:Identifier>
                        <ScaleDenominator>{matrix['scaleDenominator']}</ScaleDenominator>
                        <TopLeftCorner>{matrix['pointOfOrigin'][0]} {matrix['pointOfOrigin'][1]}</TopLeftCorner>
                        <TileWidth>{matrix['tileWidth']}</TileWidth>
                        <TileHeight>{matrix['tileHeight']}</TileHeight>
                        <MatrixWidth>{matrix['matrixWidth']}</MatrixWidth>
                        <MatrixHeight>{matrix['matrixHeight']}</MatrixHeight>
                    </TileMatrix>"""
                    tileMatrix.append(tm)

                media_type = MediaType[src.tile_type.name.lower()].value

                return self.templates.TemplateResponse(
                    "wmts.xml",
                    {
                        "request": request,
                        "tiles_endpoint": tiles_url,
                        "bounds": src.bounds,
                        "tileMatrix": tileMatrix,
                        "title": "PMTiles Archive",
                        "layer_name": "pmtiles",
                        "media_type": media_type,
                    },
                    media_type="application/xml",
                )


thatchertiler/routing.py
"""Custom routing classes."""

from typing import List, Optional

from fastapi import params
from fastapi.dependencies.utils import get_parameterless_sub_dependant
from starlette.routing import BaseRoute, Match
from typing_extensions import TypedDict


class EndpointScope(TypedDict, total=False):
    """Define endpoint."""

    # More strict version of Starlette's Scope
    # https://github.com/encode/starlette/blob/6af5c515e0a896cbf3f86ee043b88f6c24200bcf/starlette/types.py#L3
    path: str
    method: str
    type: Optional[str]  # http or websocket


def add_route_dependencies(
    routes: List[BaseRoute],
    *,
    scopes: List[EndpointScope],
    dependencies=List[params.Depends],
):
    """Add dependencies to routes.

    Allows a developer to add dependencies to a route after the route has been defined.

    """
    for route in routes:
        for scope in scopes:
            match, _ = route.matches({"type": "http", **scope})  # type: ignore
            if match != Match.FULL:
                continue

            # Mimicking how APIRoute handles dependencies:
            # https://github.com/tiangolo/fastapi/blob/1760da0efa55585c19835d81afa8ca386036c325/fastapi/routing.py#L408-L412
            for depends in dependencies[::-1]:
                route.dependant.dependencies.insert(  # type: ignore
                    0,
                    get_parameterless_sub_dependant(
                        depends=depends, path=route.path_format  # type: ignore
                    ),
                )

            # Register dependencies directly on route so that they aren't ignored if
            # the routes are later associated with an app (e.g. app.include_router(router))
            # https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/applications.py#L337-L360
            # https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/routing.py#L677-L678
            route.dependencies.extend(dependencies)  # type: ignore


thatchertiler/main.py
"""ThatcherTiler application."""

from fastapi import FastAPI
from starlette.middleware.cors import CORSMiddleware

from thatchertiler.factory import TilerFactory
from thatchertiler.settings import ApiSettings

settings = ApiSettings()


app = FastAPI(
    title="ThatcherTiler",
    openapi_url="/api",
    docs_url="/api.html",
    root_path=settings.root_path,
    description="""

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/10407788/231709578-04c9ae59-c264-4319-be9d-a70d1bd98a1f.jpg"/>
  <p align="center">ThatcherTiler: expect some features to be dropped.</p>
</p>


**Source Code**: <a href="https://github.com/developmentseed/thatchertiler" target="_blank">https://github.com/developmentseed/thatchertiler</a>

    """,
)

# Set all CORS enabled origins
if settings.cors_origins:
    app.add_middleware(
        CORSMiddleware,
        allow_origins=settings.cors_origins,
        allow_credentials=True,
        allow_methods=["GET", "OPTIONS"],
        allow_headers=["*"],
    )

tiler = TilerFactory(with_map=True, with_wmts=True)
app.include_router(tiler.router, tags=["PMTiles"])


@app.get(
    "/healthz",
    description="Health Check.",
    summary="Health Check.",
    operation_id="healthCheck",
    tags=["Health Check"],
)
def ping():
    """Health check."""
    return {"ping": "pong!"}
